### PR TITLE
Fix docker hub auth, pull image on build

### PR DESCRIPTION
**Pull request by odracci**
In this PR I fixed:
- DockerHub Auth
- Pull image defined in  `FROM` directive in Dockerfile
- Fix `sha` global variable
- Add `Login` link in the frontend
- Refactor Dockerfile to have less layers and cache `npm install`

### CODE REVIEW

File: src/docker.js [hunk 1]
```diff
@@ -132,7 +212,7 @@ docker.buildImage = function(project, tarPath, imageId, buildId, buildLogger, do
  */
 docker.tag = function(imageId, buildId, branch) {
   var deferred  = Q.defer();
-  var image     = docker.client.getImage(buildId);
+  var image     = docker.client.getImage(imageId + ':' + branch);
```
Comment (odracci) [inline src/docker.js]: This fix a race condition, when roger builds `latest` and a branch / tag at the same time, the images pushed are all `latest`

File: src/builder.js [hunk 2]
```diff
@@ -131,6 +131,7 @@ builder.build = function(project, uuid, path, gitBranch, branch, dockerOptions)
   var buildId     = imageId + ':' + branch;
   var author      = 'unknown@unknown.com';
   var now         = moment();
+  var sha         = '';
```
Comment (odracci) [inline src/builder.js]: currently `sha` is global and on concurrent builds GitHub status doesn't work, Roger share `sha` values across different builds

File: Dockerfile [hunk 3]
```diff
@@ -3,26 +3,35 @@ FROM node:8
...
+        && mkdir /tmp/roger-builds/sources
 
-RUN mkdir /tmp/roger-builds
-RUN mkdir /tmp/roger-builds/logs
-RUN mkdir /tmp/roger-builds/tars
-RUN mkdir /tmp/roger-builds/sources
+COPY ./db /db
+
+COPY ./src/client/package.json /src/src/client/
+COPY ./src/client/package-lock.json /src/src/client/
```
Comment (odino) [inline Dockerfile]: :+1:

File: src/docker.js (hunk 1 above)
Comment (odino) [reply src/docker.js]: :+1:

**Comment by odracci** [comment]
@odino I tested it with docker hub but not with a private registry

**Comment by odino** [comment]
Would you mind trying it out on a regular registry? You can use [this dumb repo](https://github.com/namshi/test-build) to try it out locally :)

**Comment by odracci** [comment]
@odino It works 😄

**Comment by odino** [comment]
I'll take your word on it :)

**Comment by odino** [comment]
BTW @odracci we have switched to Google Cloud Container Builder :) https://cloud.google.com/container-builder/docs/ still use roger here and there but most images are now built on GCP.

**Pull request merged**